inlineasm: Add inline assembler support for RV32.

agatti opened the pull request:
### Summary

This commit adds support for writing inline assembler functions when targeting a RV32IMC processor.

An almost complete test suite is also present, covering 95% of the opcodes (leaving out things like breakpoint triggers and the like).

### Testing

This was tested both on QEMU-RISCV and on an ESP32-C3 after manually enabling the inline assembler emitter in `mpconfigport.h`.

Due to the "interesting" limitations about immediates' size and registers set that vary from opcode to opcode, the code has a lot of checks for invalid input, along with (hopefully) helpful error messages as part of the exceptions it raises.  However, unless I'm doing something really wrong, I cannot seem to be able to trap those exceptions in tests.  That would be helpful for further validation of range and registers checking.

### Trade-offs and Alternatives

As mentioned in the commit message, this bit of code takes up a fair amount of .rodata space (last time I checked it was something like 4.5k) and thus it is not enabled by default unless it's running on anything that's not a microcontroller.  Since the Pico2 port has Arm inline assembler support enabled, this is also enabled for said port to maintain feature parity across MCU architectures.

Syntax-wise, there are a few things that may need a bit of explanation

* Compressed opcodes start with the `c.` prefix, which I replaced with `c_` - so for example `c.srai` turns into `c_srai`.
* When defining load/store opcodes the immediate is in the last operand, so `lw a0, 10(a1)` is defined as `lw(a0, a1, 10)`.  This is to simplify syntax parsing, as I don't think the original syntax can be parsed as such, and should be easier to grasp for folks coming from Arm assembler.
* Since writing RISC-V assembly gets painful quickly if you have to deal with inline constants, two meta opcodes have been implemented (following GAS), namely `li Rd, Imm32` and `la Rd, Label`.  The first loads the given immediate into the given register using as few bytes as possible, the second loads the address of the given label into the given register.
* `c_mov(Rd, Rs)` has been aliased to `mov(Rd, Rs)`.
* To keep code size down (most of the code in this PR is error checking anyway), there is no support for opcodes with multiple argument types (eg. `add(a0, a1)` becoming `C.ADD A0, A1` or `ADD(A0, A0, A1)` behind the scenes, or `add(a0, 100)` turning into `ADDI A0, A0, 100`).

agatti committed:
tests/inlineasm: Make room for RV32IMC inline asm tests.

Thumb/Thumb2 tests are now into their own subdirectory, as
RV32IMC-specific tests will be added as part of the RV32 inline
assembler support.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
py/misc: Add a popcount(uint32_t) implementation.

This makes the existing popcount(uint32_t) implementation found in the
RV32 emitter available to the rest of the codebase.  This version of
popcount will use intrinsic or builtin implementations if they are
available, falling back to a generic implementation if that is not the
case.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
py/emitinlinerv32: Add inline assembler support for RV32.

This commit adds support for writing inline assembler functions when
targeting a RV32IMC processor.

Given that this takes up a bit of rodata space due to its large
instruction decoding table and its extensive error messages, it is
enabled by default only on offline targets such as mpy-cross and the
qemu port.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti commented:
@dpgeorge I've resolved the conflict blocking this PR, but now the issue with Arm tests has gotten worse.  Before the qemu-arm refactoring it'd take longer than 30 seconds, but would still end.  Now it's been stuck for a couple hours already.

Am I missing something in [`e2b5402` (#15714)](https://github.com/micropython/micropython/pull/15714/commits/e2b540231db68e24f80feeebe0cd3030b88e0517) when moving files around?  I can't see what's wrong, unless there's some extra logic elsewhere I haven't updated.

dpgeorge commented:
> but now the issue with Arm tests has gotten worse. Before the qemu-arm refactoring it'd take longer than 30 seconds, but would still end. Now it's been stuck for a couple hours already.

It looks like this issue with CI is resolved now?  The qemu-arm CI job now passes without issue, it seems.

agatti commented:
Yes, that issue was indeed resolved - however the inlineasm tests for RV32 won't be picked up until #15784 is merged (and until this PR gets rebased once more).

agatti committed:
qemu/main: Make GC heap size configurable on a per-arch basis.

In certain circumstances depending on the code size, the
`deflate_decompress` test fails on both ARM and RV32 with a memory
allocation failure error.  The issue is mitigated by having a larger GC
heap, in this case around 20 KBytes more than the original 100 KBytes
default.

This commit makes the GC heap size configurable on a per-arch basis, with
both ARM and RV32 using the enlarged 120 KBytes heap.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti commented:
Here's the rebased version taking the merged `qemu` port into account.  I've run the tests locally in a VM being as close as possible to the CI environment, and on an ESP32C3 board.

Running tests for `MPS2_AN385`, `SABRELITE`, and `VIRT_RV32` qemu boards work fine, but `MICROBIT` and `NETDUINO2` targets do not work at all (even without this PR).  Are those boards meant to be compile-only?

`MICROBIT` test run:

<details>

```
dev@mpy-dev-rv32:/micropython/ports/qemu$ make BOARD=MICROBIT test V=1
python3 ../../py/makeversionhdr.py build-MICROBIT/genhdr/mpversion.h
test -e "../../lib/micropython-lib/README.md" || (echo -e "\033[1;31mError: micropython-lib submodule is not initialized.\033[0m Run 'make submodules'"; false)
python3 ../../tools/makemanifest.py -o build-MICROBIT/frozen_content.c -v "PORT_DIR=/micropython/ports/qemu" -v "MPY_DIR=../.." -v "BOARD_DIR=" -v "MPY_LIB_DIR=../../lib/micropython-lib" -b "build-MICROBIT" -f"-march=armv7m" --mpy-tool-flags="" "freeze('test-frzmpy', ('frozen_asm_thumb.py', 'frozen_const.py', 'frozen_viper.py', 'native_frozen_align.py'))"
cd ../../tests && ./run-tests.py --target qemu --device execpty:"qemu-system-arm -global nrf51-soc.flash-size=1048576 -global nrf51-soc.sram-size=262144 -machine microbit -nographic -monitor null -semihosting  -serial pty -kernel ../ports/qemu/build-MICROBIT/firmware.elf" --exclude 'inlineasm/rv32imc' 
# HANGS HERE
```

</details>

`NETDUINO2` test run:

<details>

```
dev@mpy-dev-rv32:/micropython/ports/qemu$ make BOARD=NETDUINO2 test V=1
python3 ../../py/makeversionhdr.py build-NETDUINO2/genhdr/mpversion.h
test -e "../../lib/micropython-lib/README.md" || (echo -e "\033[1;31mError: micropython-lib submodule is not initialized.\033[0m Run 'make submodules'"; false)
python3 ../../tools/makemanifest.py -o build-NETDUINO2/frozen_content.c -v "PORT_DIR=/micropython/ports/qemu" -v "MPY_DIR=../.." -v "BOARD_DIR=" -v "MPY_LIB_DIR=../../lib/micropython-lib" -b "build-NETDUINO2" -f"-march=armv7m" --mpy-tool-flags="" "freeze('test-frzmpy', ('frozen_asm_thumb.py', 'frozen_const.py', 'frozen_viper.py', 'native_frozen_align.py'))"
cd ../../tests && ./run-tests.py --target qemu --device execpty:"qemu-system-arm -machine netduino2 -nographic -monitor null -semihosting  -serial pty -kernel ../ports/qemu/build-NETDUINO2/firmware.elf" --exclude 'inlineasm/rv32imc' 
b'MicroPython v1.24.0-preview.297.g34fbf1d55 on 2024-09-06; netduino2 with STM32\r\nType "help()" for more information.\r\n>>> '
Traceback (most recent call last):
  File "/micropython/tests/./run-tests.py", line 1188, in <module>
    main()
  File "/micropython/tests/./run-tests.py", line 1084, in main
    pyb.enter_raw_repl()
  File "/micropython/tools/pyboard.py", line 365, in enter_raw_repl
    raise PyboardError("could not enter raw repl")
pyboard.PyboardError: could not enter raw repl
make: *** [Makefile:169: test] Error 1
```

</details>

dpgeorge commented:
> but `MICROBIT` and `NETDUINO2` targets do not work at all (even without this PR). Are those boards meant to be compile-only?

You are right, these boards don't run the tests on master.

Looking into it it seems the problems are:
- NETDUINO2: requires a `time.sleep(0.1)` in `tools/pyboard.py`, `ProcessPtyToTerminal::__init__`, just before creating the `serial.Serial(pty, ...)` instance.  That gives qemu enough time to start up before trying to access the serial port.
- MICROBIT: probably needs the same patch as NETDUINO2, but then also locks up running the `tests/feature_check/native_check.py`.  Solution would be to either debug why it crashes on that and fix it, or just disable the native emitter on this board.

This is unrelated to this PR and so out of scope.  But feel free to fix it if you like :smile:

agatti commented:
@dpgeorge fyi, MICROBIT crashes when accessing the function table, which doesn't make much sense.

```
...
(gdb) 
0x20000f74 in gc_heap ()
=> 0x20000f74 <gc_heap+3192>:   00 a8   add     r0, sp, #0
(gdb) 
0x20000f76 in gc_heap ()
=> 0x20000f76 <gc_heap+3194>:   d7 f8 b4 40     ldr.w   r4, [r7, #180]  ; 0xb4
(gdb) print/a $r7
$3 = 0x28fe8 <mp_fun_table>
(gdb) ni
0x0001ff8a in HardFault_Handler ()
=> 0x0001ff8a <HardFault_Handler+0>:    03 20   movs    r0, #3
(gdb) x/4a mp_fun_table+180
0x2909c <mp_fun_table+180>:     0x1b689 <mp_setup_code_state_native>    0x1bb9b <mp_small_int_floor_divide>     0x1bb7d <mp_small_int_modulo>   0xe23d <mp_native_yield_from>
```

agatti committed:
rp2/mpconfigport: Enable RV32 inline assembly support.

This commit enables by default inline assembly support for the RP2 target
when it is operating in RISC-V mode.  This brings the feature set when in
RISC-V mode to parity with what's available in ARM mode.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
tests/run-tests.py: Detect inlineasm support and add tests if needed.

This commit implements a method to detect at runtime if inline assembler
support is enabled, and if so which platform it targets.

This allows clean test runs even on modified version of ARM-based ports
where inline assembler support is disabled, running inline assembler tests
on ports that have such feature not enabled by default and manually
enabled, and allows to always run the correct inlineasm tests for ports
that support more than one architecture (esp32, qemu, rp2).

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti commented:
This should now work with the changes to the test runner made after 1.24.0.

This PR also makes the test runner able to detect if inline assembler support is enabled and if so for which architecture.  If so, the `tests/inlineasm/$INLINEASM_ARCH/` directory is automatically added to the test suite.  On esp32, qemu, and rp2 it will pick up the correct test files according to what arch is currently in use (allowing for future inlineasm additions like RV64 and other architectures/variants).

agatti commented:
I'm wondering on what's the policy for new test files after the introduction of `unittest`.  

Should the test files contained in this PR moved to that framework (as in, is it a blocking issue) or is it something that can be done at a later stage (not necessarily by myself, either)?

dpgeorge commented:
> I'm wondering on what's the policy for new test files after the introduction of `unittest`.

Using unittest is only for convenience, if it makes writing the test easier.  So you don't need to change anything in this PR.

dpgeorge commented on ports/qemu/Makefile:
```diff
@@ -19,16 +19,20 @@ QSTR_DEFS = qstrdefsport.h
 endif
 ifeq ($(QEMU_ARCH),riscv32)
-FROZEN_MANIFEST ?= "require('unittest'); freeze('test-frzmpy', ('frozen_const.py', 'frozen_viper.py', 'native_frozen_align.py'))"
+QEMU_GC_HEAP_SIZE ?= 120
```
Can we just increase the heap size to 120k for all boards/archs?  That would keep it simple (if it's possible to do, if the boards all have enough RAM for that, even though it's all virtual).

I would prefer to keep it simple and use the one size.  But if you think it's needed (or a much better idea) to have this configurable, then I suggest a different name for the variable.  The prefix `QEMU_` indicates that it's a qemu-related config option, when it's actually a MicroPython option.  So maybe just `GC_HEAP_SIZE`, or `MICROPY_GC_HEAP_SIZE`.

agatti replied:
Sharing the same default for both Arm and RV32 could work but then that value would need to be changed once more for #15931, to a value that may not reflect the reality of the physical boards the code under test will run on.  Unless RV64 ends up with a different default, and at that point it's pretty much the same as having a per-arch default.

Besides fixing the RV32 tests (for which I still have no idea why that happens), I've extensively used this feature to see how my code behaves in low-memory conditions, due to a simulated testing scenario, or due to a fragmented heap after long periods of (simulated) execution time.  I found it much faster to script a command line argument to `make` than messing with `CFLAGS`.

W.r.t. the prefix, you've got valid points, but on the other hand this flag is specific to the QEMU test runner.  I'd say that keeping the prefix sort of makes it more explicit that it is strictly a testing-only configuration flag, but I'm not 100% against any changes on that front.

agatti replied:
No worries, I missed the bit about those other two ports - I rarely interact with minimal or with zephyr, if at all.

I've updated the user-facing name to `MICROPY_HEAP_SIZE` in both the code and the port's README file.

dpgeorge commented on ports/qemu/main.c:
```diff
@@ -34,7 +34,11 @@
+#error MICROPY_HEAP_SIZE must be a positive integer.
+#endif
+
+#define HEAP_SIZE (MICROPY_HEAP_SIZE * 1024)
```
The minimal and zephyr ports specify `MICROPY_HEAP_SIZE` in bytes, so it should also be bytes here, not kibytes.

agatti replied:
This should be handled in the latest changes I pushed.

dpgeorge commented:
> * Compressed opcodes start with the `c.` prefix, which I replaced with `c_` - so for example `c.srai` turns into `c_srai`

It would be possible to allow `c.` as the prefix, because it's valid Python syntax.  It would just be a little bit of extra code to identify in the emitter.

Do you think it's worth trying to match the syntax more closely here, and use `c.` as the prefix?

Note that the Thumb emitter uses `_w` suffix for wide instructions which are `.w` in ARM syntax, eg `beq.w label` is `beq_w(label)` in MicroPython.  So for consistency maybe just worth sticking with `c_` for RISC-V. 

> * When defining load/store opcodes the immediate is in the last operand, so `lw a0, 10(a1)` is defined as `lw(a0, a1, 10)`.  This is to simplify syntax parsing, as I don't think the original syntax can be parsed as such, and should be easier to grasp for folks coming from Arm assembler.

It would be possible to support `lw(a0, 10(a1))`, because that's valid Python syntax.

Note that the Thumb emitter uses eg `ldr(r0, [r1, 12])` for specifying an offset to load/store, so it matches actual Thumb assembler (see `py/emitinlinethum.c:get_addr_arg`).  So I'm thinking it would be good to match RISC-V assembler as well and support `10(a1)` syntax.  It would be quite simple to do, just make a helper function like `get_addr_arg` that matches the `RULE_atom_expr_normal` with the second element of that node being a `RULE_trailer_paren`.

> * `c_mov(Rd, Rs)` has been aliased to `mov(Rd, Rs)`.

Do you mean that using `mov()` it will emit a `c_mov()` instruction?  (That makes more sense to me than the opposite.)


> * To keep code size down (most of the code in this PR is error checking anyway), there is no support for opcodes with multiple argument types

That sounds very reasonable.

---------

> the code has a lot of checks for invalid input, along with (hopefully) helpful error messages as part of the exceptions it raises
> However, unless I'm doing something really wrong, I cannot seem to be able to trap those exceptions in tests. That would be helpful for further validation of range and registers checking.

You can probably get these tests working by using `exec("<bad inline code>")`, eg:
```py
def test_asm(code):
    try:
        exec(code)
    except SyntaxError:
        print(...)

test_asm("""
@micropython.asm_rv32
def f(a0):
    mov(a100, a200)
""")
```

agatti commented:
> > * Compressed opcodes start with the `c.` prefix [...]

> Do you think it's worth trying to match the syntax more closely here, and use `c.` as the prefix?

Given that the Thumb emitter doesn't use dots in opcode names, as you point out, I'd say that leaving `c_` as a prefix right now is a decent compromise to get this out of the door sooner.

Maybe the dot notation can be introduced for both RISC-V and Thumb in v2, as some breakage may be expected then.

> It would be possible to support `lw(a0, 10(a1))`, because that's valid Python syntax.

Is it?  I thought integers were validated in their own specialised way in the code to speed things up - ints do not implement `__call__` so I expected that having an open parenthesis right after an int would trigger an error before the inline assembler parser would take control of the evaluation.  Good to know, I'll see what I can do for that, and thanks for the hints about the specialised parsing.

> Do you mean that using `mov()` it will emit a `c_mov()` instruction? (That makes more sense to me than the opposite.)

Indeed, `mov` isn't even an official RISC-V opcode - sorry for the unwanted confusion (it made sense in Italian :)).

> You can probably get these tests working by using `exec("<bad inline code>")`, eg: [...]

I'll give that a shot, thanks!

agatti commented:
`RULE_atom_expr_normal` unfortunately fails on expressions like `lw(a0, -4(sp))`, in that case it matches `RULE_factor_2`.  I'll work around that somehow.

agatti commented:
This is still missing the syntax error tests - I just pushed the changes so the C code changes can be reviewed as I'm not 100% sure whether the new parse node handling code is correct.

dpgeorge commented on py/emitinlinerv32.c:
```diff
@@ -0,0 +1,774 @@
+    assert(offset_node != NULL && "Offset node pointer is NULL.");
+    assert(negative != NULL && "Negative pointer is NULL.");
+
+    if (!MP_PARSE_NODE_IS_STRUCT_KIND(node, PN_atom_expr_normal) && !MP_PARSE_NODE_IS_STRUCT_KIND(node, PN_factor_2)) {
```
This validation logic can be simplified, in pseudo code:
```
if MP_PARSE_NODE_IS_STRUCT_KIND(node, PN_factor_2):
    if MP_PARSE_NODE_IS_TOKEN_KIND(node_struct->nodes[0], MP_TOKEN_OP_MINUS):
        *negative = true;
    elif MP_PARSE_NODE_IS_TOKEN_KIND(node_struct->nodes[0], MP_TOKEN_OP_PLUS):
        // pass
    else:
        goto invalid;
    node = node_struct->nodes[1]

if !MP_PARSE_NODE_IS_STRUCT_KIND(node, PN_atom_expr_normal):
    goto invalid
```

Note that `PN_factor_2` and `PN_atom_expr_normal` always have 2 elements so there's no need to check that.

agatti replied:
Thanks, I ended up doing something similar, but I left the elements count check.  I'll take that out.

Right now I'm cleaning up the range checks, as negative offsets are still seen as positive due to the offset node being without its minus sign and you can imagine the rest.  Right now there are lots of repeated code for the special cases and I'm merging together.

dpgeorge commented:
When an rv32 inline-asm function is entered into, what registers are automatically saved?  Is there anything else done in the prelude?

With Thumb inline-asm, a few registers are saved/restored on function entry/exit.  This was arguably not a good choice, it might have been a better choice to have a very minimal prelude for inline-asm function.

agatti commented:
> When an rv32 inline-asm function is entered into, what registers are automatically saved? Is there anything else done in the prelude?

The prelude saves the function table pointer and the three local registers.  The reason behind that is purely because they're stored in S-registers and those registers must be restored on exit (in other words, A0-A7 and T0-T6 can be trashed as much as you want, but S0-S11 must be preserved).

In theory T-registers or even the A4 to A7 could be have been used for those, but then half of the compressed opcodes wouldn't be able to access them.

In fact, the emitter just calls `asm_rv32_entry`/`asm_rv32_exit`	and one of the changes to `py/asmrv32.c` present in this PR is to save an internal temporary S-register only whenever there's an emitted function call (that's required to shorten most function calls to just 2 opcodes/4 bytes).

For inline assembly code it's up to the user to keep track of either RA or S0-S11, but that's expected of the user in the first place no matter the assembler used.

agatti commented:
OK, this should be it I guess.  I wish the offset-to-register syntax support wasn't bolted on this way, but it works :)  That can be cleaned up for v2 anyway.

dpgeorge commented:
> the offset-to-register syntax support wasn't bolted on this way

What do you mean, that you'd prefer it was designed it from the start?

agatti commented:
Indeed, everything else fits in nicely without special cases and exceptions.  It's more of a personal preference, really, but re-design everything to cover that case in a generic way as well would take a bit more time and delay this PR for the 1.25.0 milestone.

Anyway, I'll rework that for v2 when adding dotted opcodes support - in the end it works and the code isn't *that* bad, but it could be better for sure :)

dpgeorge commented:
OK, sounds fine to me.  I'm glad that the syntax `10(a0)` is now supported, I think that's a good improvement.

dpgeorge commented:
> The prelude saves the function table pointer and the three local registers.

Just to be clear, this is necessary for @native/@viper code generation.  But for inline-rv32-asm it's not strictly needed, but done just because inline-rv32-asm is using `asm_rv32_entry`/`asm_rv32_exit` for convenience.  Is that true?

Would it be possible for inline-rv32-asm to use a different entry/exit set of functions and not save any registers automatically?  As you say, it's up to the assembler user to keep track of and restore needed registers.

agatti commented:
Done.  Maybe when more targets will support `cm.push`/`cm.popret` opcodes (right now only the pico2 does) it can be feasible to automatically save/restore RA + S0..S11 with an explicit opt-out via an argument to the decorator (that's 2 opcodes/4 bytes vs ~60 bytes/~25 opcodes).

agatti commented:
Apologies for the extra push, it's just a fix for a minor issue I found when reviewing the code for the last time (the check for negative values' range didn't take into account that the negative flag was a pointer), plus some commented test code was removed.

dpgeorge commented on py/asmrv32.c:
```diff
@@ -126,7 +100,6 @@ static void split_immediate(mp_int_t immediate, mp_uint_t *upper, mp_uint_t *low
     // Turn the lower half from unsigned to signed.
     if ((*lower & 0x800) != 0) {
         *upper += 0x1000;
-        *lower -= 0x1000;
```
Is removing this line intentional?  Is it a bug fix or refactoring?

agatti replied:
More like both a bug fix and a refactoring.  `*lower` would contain the lowest 12 bits of a 32 bits immediate, so applying a further offset isn't correct in that case even if it leaves the lowest part alone; however the variable has its upper 20 bits masked out on every usage so the extra offset didn't really matter.

dpgeorge replied:
OK

dpgeorge commented on tests/ports/qemu/asm_test.py:
```diff
@@ -1,4 +1,13 @@
+elif arch == 11:
+    import frozen_asm_rv32 as frozen_asm
+else:
+    raise ImportError
```
Would it be simpler and equivalent to do:
```py
try:
    import frozen_asm_thumb as frozen_asm
except ImportError:
    try:
        import frozen_asm_rv32 as frozen_asm
    except ImportError:
        raise ImportError
```
?

Then you don't need to know about the different arch values.

agatti replied:
Makes sense.  This should be fixed in the latest iteration of this PR.

dpgeorge commented on py/emitinlinerv32.c:
```diff
@@ -0,0 +1,818 @@
+    { MP_QSTR_li,         opcode_li,                 CALL_RI,  R,   0, I,   0,  N,  0, 0xFFFFFFFF, 0xFFFFFFFF, 0x00000000 },
+    { MP_QSTR_lui,        asm_rv32_opcode_lui,       CALL_RI,  R,   0, I,   12, N,  0, 0xFFFFFFFF, 0xFFFFF000, 0x00000000 },
+    { MP_QSTR_lw,         asm_rv32_opcode_lw,        CALL_RIR, R,   0, I,   0,  R,  0, 0xFFFFFFFF, 0x00000FFF, 0xFFFFFFFF },
+    { MP_QSTR_mov,        asm_rv32_opcode_cmv,       CALL_RR,  R,   0, R,   0,  N,  0, 0xFFFFFFFF, 0xFFFFFFFF, 0x00000000 },
```
I suggest calling this pseudo/convenience opcode `mv`.

That matches GAS (as far as I can tell, which aliases `mv` to `c.mv`) and also matches the fact that `c.mv` is called `c.mv` rather than `c.mov`.

agatti replied:
I reckoned GAS used `mov`, but I've always used `c.mv` or `add(rd, rm, zero)` myself depending on my needs.  Anyway, I've just checked and GAS indeed uses `mv`, thanks for spotting that!  

This should be fixed now.

dpgeorge reviewed (commented):
This is looking very good now.  I particularly like how the tests suite can now auto-detect the inline-asm capability and arch, that's nice!

I have a few final comments and then it should be good to go in.

dpgeorge reviewed (approved):
Tested on `RPI_PICO2` in RISC-V mode.  Works well.

dpgeorge commented:
@agatti thanks for your efforts on this, it's a very nice feature to have, especially because RISC-V will become more and more prevalent in MCUs in the future.

BTW, if you want to have a go at reducing the code size, you should be able to cut off at least 1k with the following tricks:
- use `qstr_short_t` for the type of `REGISTERS_QSTR_TABLE`
- for the `argumentX_mask` variables of `opcode_t`, make them indirect enums into a table of masks (because there are only a limited set of possible masks), eg `enum { MASK_00000000, MASK_00000FFF, MASK_FFFFFFFF, ...}` then access via `MASK_TABLE[opcode->argument1_mask]`.
- combined with the above, use `qstr_short_t` for the `qstring` entry in `opcode_t` and 5-bit fields for `argumentX_mask`, which should allow fitting the qstr and the 3 argument masks in 32 bits

agatti commented:
>     * use `qstr_short_t` for the type of `REGISTERS_QSTR_TABLE`

Oh, I don't think I've ever seen `qstr_short_t` being used before.  So I must assume there's a hard limit of 64k qstrs?

>     * for the `argumentX_mask` variables of `opcode_t`, make them indirect enums into a table of masks [...].

That's nice - wish I've thought of that :)  This will end up as a PR after 1.25.0 is out, so it can be featured in a later version along with eventual fixes for bugs that may be found in the meantime.

All things considered, reducing the footprint by 30% or more may probably make this a candidate for being enabled by default on ESP32Cx targets too.  I can probably shorten some error messages' text and coalesce some messages as well to further reduce the space taken up by this feature, I'll see what can be done on that.

Anyway, thanks for the help on this - glad it's been merged!

dpgeorge commented:
> I can probably shorten some error messages' text and coalesce some messages as well to further reduce the space taken up by this feature,

Note that there are configurable error message levels:
```
// Exception messages are removed
#define MICROPY_ERROR_REPORTING_NONE     (0)                                   
// Exception messages are short static strings                                 
#define MICROPY_ERROR_REPORTING_TERSE    (1)                                 
// Exception messages provide basic error details     
#define MICROPY_ERROR_REPORTING_NORMAL   (2)                                 
// Exception messages provide full info, e.g. object names                       
#define MICROPY_ERROR_REPORTING_DETAILED (3)                                                                                
```
The default for most ports is NORMAL.  It's possible to provide different error messages for the different levels.

But actually I think what you have done already is pretty good, it's a much better user experience with more detailed error messages for tricky things like inline assembler.  Maybe it could be tweaked to reduce size by a little, eg I just noticed a few messages end in `.` which should be removed.
